samples/native_training/main.c

```diff
@@ -0,0 +1,236 @@
+      session, iree_make_cstring_view("module.forward"), &call));
+
+  // Allocate buffers in device-local memory so that if the device has an
+  // independent address space they live on the fast side of the fence.
+  iree_hal_dim_t shape_w[1] = {IREE_ARRAYSIZE(state->w)};
+  iree_hal_dim_t shape_b[0] = {};
+  iree_hal_dim_t shape_X[2] = {1, IREE_ARRAYSIZE(state->X)};
+  iree_hal_dim_t shape_y[1] = {IREE_ARRAYSIZE(state->y)};
+  iree_hal_buffer_view_t* arg0 = NULL;
+  iree_hal_buffer_view_t* arg1 = NULL;
+  iree_hal_buffer_view_t* arg2 = NULL;
+  iree_hal_buffer_view_t* arg3 = NULL;
+  IREE_RETURN_IF_ERROR(iree_hal_buffer_view_allocate_buffer(
+      iree_runtime_session_device_allocator(session), IREE_ARRAYSIZE(shape_w),
+      shape_w, IREE_HAL_ELEMENT_TYPE_FLOAT_32,
+      IREE_HAL_ENCODING_TYPE_DENSE_ROW_MAJOR,
+      (iree_hal_buffer_params_t){
+          .usage = IREE_HAL_BUFFER_USAGE_DEFAULT,
+          .type = IREE_HAL_MEMORY_TYPE_DEVICE_LOCAL,
+      },
+      iree_make_const_byte_span(state->w, sizeof(state->w)), &arg0));
+  IREE_RETURN_IF_ERROR(iree_hal_buffer_view_allocate_buffer(
+      iree_runtime_session_device_allocator(session), IREE_ARRAYSIZE(shape_b),
+      shape_b, IREE_HAL_ELEMENT_TYPE_FLOAT_32,
+      IREE_HAL_ENCODING_TYPE_DENSE_ROW_MAJOR,
+      (iree_hal_buffer_params_t){
+          .usage = IREE_HAL_BUFFER_USAGE_DEFAULT,
+          .type = IREE_HAL_MEMORY_TYPE_DEVICE_LOCAL,
+      },
+      iree_make_const_byte_span(state->b, sizeof(state->b)), &arg1));
+  IREE_RETURN_IF_ERROR(iree_hal_buffer_view_allocate_buffer(
+      iree_runtime_session_device_allocator(session), IREE_ARRAYSIZE(shape_X),
+      shape_X, IREE_HAL_ELEMENT_TYPE_FLOAT_32,
+      IREE_HAL_ENCODING_TYPE_DENSE_ROW_MAJOR,
+      (iree_hal_buffer_params_t){
+          .usage = IREE_HAL_BUFFER_USAGE_DEFAULT,
+          .type = IREE_HAL_MEMORY_TYPE_DEVICE_LOCAL,
+      },
+      iree_make_const_byte_span(state->X, sizeof(state->X)), &arg2));
+  IREE_RETURN_IF_ERROR(iree_hal_buffer_view_allocate_buffer(
+      iree_runtime_session_device_allocator(session), IREE_ARRAYSIZE(shape_y),
+      shape_y, IREE_HAL_ELEMENT_TYPE_FLOAT_32,
+      IREE_HAL_ENCODING_TYPE_DENSE_ROW_MAJOR,
+      (iree_hal_buffer_params_t){
+          .usage = IREE_HAL_BUFFER_USAGE_DEFAULT,
+          .type = IREE_HAL_MEMORY_TYPE_DEVICE_LOCAL,
+      },
+      iree_make_const_byte_span(state->y, sizeof(state->y)), &arg3));
+
+  // Setup call inputs with our buffers.
+  IREE_RETURN_IF_ERROR(
+      iree_runtime_call_inputs_push_back_buffer_view(&call, arg0));
+  iree_hal_buffer_view_release(arg0);
+  IREE_RETURN_IF_ERROR(
+      iree_runtime_call_inputs_push_back_buffer_view(&call, arg1));
+  iree_hal_buffer_view_release(arg1);
+  IREE_RETURN_IF_ERROR(
+      iree_runtime_call_inputs_push_back_buffer_view(&call, arg2));
+  iree_hal_buffer_view_release(arg2);
+  IREE_RETURN_IF_ERROR(
+      iree_runtime_call_inputs_push_back_buffer_view(&call, arg3));
+  iree_hal_buffer_view_release(arg3);
```

Comment:
These arg0-arg3 buffer views should still be released if any of the allocations above fail. Might want to change this part of the function (lines 34-93) to the `if (iree_status_is_ok(status)) {` style.

Maybe @benvanik has other suggestions on how to structure this C code.

Reply:
Ben has this comment in another sample :P

```
// NOTE: CHECKs are dangerous but this is a sample; a real application would
// want to handle errors gracefully. We know in this constrained case that
// these won't fail unless something is catastrophically wrong (out of memory,
// solar flares, etc).
```

Then he uses `IREE_CHECK_OK` instead of `IREE_RETURN_IF_ERROR`: https://github.com/iree-org/iree/pull/11326/commits/eefc3b042f10b1fdc66bd3d54dda0c3b638ad3ab

Reply:
Went with your first suggested approach, thanks for catching.